File: 21_merge_two_sorted_list/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):

    def merge_two_sorted_lists(self, l1, l2):
        head = ListNode(0);
        p = head;

        while l1 is not None and l2 is not None:
            num = 0
            if l1.val < l2.val:
                num = l1.val
                l1 = l1.next
            else:
                num = l2.val
                l2 = l2.next
            p.next = ListNode(num)
            p = p.next

        p.next = l1 if l1 is not None else l2
        return head.next


def print_list(l):
    p = l
    while p is not None:
        print(p.val)
        p = p.next
    

if __name__ == '__main__':
    l1 = ListNode(1)
    l1_1 = ListNode(2)
    l1.next = l1_1
    l1_2 = ListNode(4)
    l1_1.next = l1_2

    l2 = ListNode(1)
    l2_1 = ListNode(3)
    l2.next = l2_1
    l2_2 = ListNode(4)
    l2_1.next = l2_2

    s = Solution()
    print_list(s.merge_two_sorted_lists(l1, l2))

File: 21_merge_two_sorted_list/go/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoSortedLists(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{Val: 0}
	p := head

	var num int
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			num = l1.Val
			l1 = l1.Next
		} else {
			num = l2.Val
			l2 = l2.Next
		}
		p.Next = &ListNode{Val: num}
		p = p.Next
	}
	if l1 != nil {
		p.Next = l1
	} else {
		p.Next = l2
	}
	return head.Next
}

func printList(l *ListNode) {
	p := l
	for p != nil {
		fmt.Printf("%d ", p.Val)
		p = p.Next
	}
	fmt.Println()
}

func main() {
	l1 := &ListNode{Val: 1}
	l1_1 := &ListNode{Val: 2}
	l1.Next = l1_1
	l1_2 := &ListNode{Val: 4}
	l1_1.Next = l1_2

	l2 := &ListNode{Val: 1}
	l2_1 := &ListNode{Val: 3}
	l2.Next = l2_1
	l2_2 := &ListNode{Val: 4}
	l2_1.Next = l2_2

	printList(mergeTwoSortedLists(l1, l2))
}
